src/mlpack/methods/mean_shift/mean_shift.hpp

```diff
@@ -0,0 +1,116 @@
+/**
+ * @file mean_shift.hpp
+ * @author Shangtong Zhang
+ *
+ * Mean Shift clustering
+ */
+
+#ifndef __MLPACK_METHODS_MEAN_SHIFT_MEAN_SHIFT_HPP
+#define __MLPACK_METHODS_MEAN_SHIFT_MEAN_SHIFT_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/kernels/gaussian_kernel.hpp>
+
+namespace mlpack {
+namespace meanshift /** Mean Shift clustering. */ {
+
+/**
+ * This class implements Mean Shift clustering.
+ * For each point in dataset, apply mean shift algorithm until maximum 
+ * iterations or convergence. 
+ * Then remove duplicate centroids.
+ * 
+ * A simple example of how to run Mean Shift clustering is shown below.
+ *
+ * @code
+ * extern arma::mat data; // Dataset we want to run Mean Shift on.
+ * arma::Col<size_t> assignments; // Cluster assignments.
+ * arma::mat centroids; // Cluster centroids.
+ *
+ * MeanShift<arma::mat, kernel::GaussianKernel> meanShift();
+ * meanShift.Cluster(dataset, assignments, centroids);
+ * @endcode
+ *
+ * @tparam KernelType the kernel to use.
+ */
+  
+template<typename KernelType = kernel::GaussianKernel,
+         typename MatType = arma::mat>
+class MeanShift
+{
+ public:
+  /**
+   * Create a Mean Shift object and set the parameters which Mean Shift
+   * will be run with.
+   *
+   * @param duplicateThresh If distance of two centroids is less than it, one will be removed.
+   * @param maxIterations Maximum number of iterations allowed before giving up
+   * @param stopThresh If the 2-norm of the mean shift vector is less than stopThresh, 
+   *        iterations will terminate.
+   * @param kernel Optional KernelType object.
+   */
+  MeanShift(const double duplicateThresh = 1.0,
+            const size_t maxIterations = 1000,
+            const double stopThresh = 1e-3,
+            const KernelType kernel = KernelType());
```

Comment:
I don't think that the duplicate threshold should have a default, because the value that should be used is completely dependent on the dataset.  I don't mind putting in some basic logic to estimate a threshold value if the user does not specify one, but a default of 1.0 is a bad idea, in my opinion.  (If you think I'm wrong, please feel free to provide a rebuttal.)